test/common/router/rds_impl_test.cc

```diff
@@ -278,6 +279,39 @@ TEST_F(RdsImplTest, FailureSubscription) {
   rds_callbacks_->onConfigUpdateFailed(Envoy::Config::ConfigUpdateFailureReason::FetchTimedout, {});
 }
 
+// Verifies that a queued up request for a virtual host update doesn't crash if
+// RdsRouteConfigProvider is deallocated
+TEST_F(RdsImplTest, VirtualHostUpdateWhenProviderHasBeenDeallocated) {
+  const std::string rds_config = R"EOF(
+rds:
+  route_config_name: my_route
+  config_source:
+    api_config_source:
+      api_type: GRPC
+      grpc_services:
+        envoy_grpc:
+          cluster_name: xds_cluster
+)EOF";
+
+  Event::PostCb post_cb;
+  testing::NiceMock<Event::MockDispatcher> local_thread_dispatcher;
+  {
+    auto rds = RouteConfigProviderUtil::create(
+        parseHttpConnectionManagerFromYaml(rds_config), server_factory_context_,
+        validation_visitor_, outer_init_manager_, "foo.", *route_config_provider_manager_);
+
+    EXPECT_CALL(server_factory_context_.dispatcher_, post(_)).WillOnce(SaveArg<0>(&post_cb));
+    rds->requestVirtualHostsUpdate("testing", local_thread_dispatcher,
+                                   std::make_shared<Http::RouteConfigUpdatedCallback>(
+                                       Http::RouteConfigUpdatedCallback([](bool) -> void {})));
```

Comment:
nit: might worth check if the lambda is actually not called. 

The dangling reference may not lead to an Exception but some segv error.